design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
+
+## Service Lifecycle
+
+The new `Service` construct differs a bit from other CDK L2+ constructs. Most CDK constructs are fully assembled in the constructor and are then immutable. The `Service` class is not like this. The constructor does nothing but initialize an empty list of extensions. Users add extensions to this list, to define their intention about what should be built later on. The service constructs do not actually exist at this point. The only thing that exists is a list of user specified intentions about what constructs they do want to exist.
+
+The service's constructs are instantiated once the service is added to an environment. At that point all the extensions are executed to build out the actual CDK resources. From this point on the service constructs are available and immutable.
+
+## Extension Implemention
+
+Each extension is a collection of functions which are executed by the `Service` class when the service is added to an environment. Functions are executed in the order that the extensions were added to the service. Some functions are mutating hooks which modify service or task definition properties prior to these constructs being created. Other functions are called after each construct has been created so that the extension can make use of the construct.
+
+The hooks for an extension are as follows:
+
+- `addHooks()` - This hook is called after all the extensions are added, but before anything else has been run. It gives each extension a chance to do some inspection of the overall service and see what other extensions have been added. Some addons may want to register hooks on the other addons to modify them. For example the Firelens addon wants to be able to modify the settings of the application container to route logs through Firelens.
+- `mutateTaskDefinitionProps()` - This is a mutating hook which is passed the `ecs.TaskDefinitionProps` for a TaskDefinition. This allows the extension to make modifications to the task definition, such as App Mesh extension modifying the proxy settings for the task.
+- `useTaskDefinition()` - After the task definition is created, this hook is passed the actual `TaskDefinition` construct that was created. This allows the extension to add containers to the task, modify the task definition's IAM role, etc.
+- `bakeContainerDependencies()` - Once all extensions have added their containers each addon is given a chance to modify its container's `dependsOn` settings. Extensions need to check and see what other extensions were enabled, and decide whether their container needs to wait on another container to start first.
```

Comment:
Would `resolveContainerDependencies` be a better descriptor here?

Reply:
Sure that sounds good to me